src/main/java/org/mskcc/cbio/oncokb/web/rest/ApiProxy.java

```diff
@@ -69,41 +72,57 @@ public String proxy(@RequestBody(required = false) String body, HttpMethod metho
         List<String> tokenUsageCheckList = Arrays.stream(applicationProperties.getTokenUsageCheck().split(",")).map(api -> api.trim()).filter(api -> !api.isEmpty()).collect(Collectors.toList());
         Optional<String> needsToBeRecorded = tokenUsageCheckList.stream().filter(api -> request.getRequestURI().startsWith(api)).findFirst();
         if (needsToBeRecorded.isPresent()) {
-            Optional<String> userOptional = SecurityUtils.getCurrentUserLogin();
-            if (userOptional.isPresent()) {
-                List<String> tokenUsageCheckWhitelist = Arrays.stream(applicationProperties.getTokenUsageCheckWhitelist().split(",")).map(api -> api.trim()).filter(api -> !api.isEmpty()).collect(Collectors.toList());
-
-                Optional<User> user = userService.getUserWithAuthoritiesByLogin(userOptional.get());
-                Optional<UUID> uuidOptional = SecurityUtils.getCurrentUserToken();
-                if (user.isPresent() &&
-                    user.get().getAuthorities().stream().filter(authority -> authority.getName().equalsIgnoreCase(AuthoritiesConstants.ADMIN)).count() == 0 &&
-                    !tokenUsageCheckWhitelist.contains(user.get().getLogin())) {
-                    List<Token> tokenList = tokenProvider.getUserTokens(user.get());
-                    tokenList.forEach(token -> {
-                        token.setCurrentUsage(token.getCurrentUsage() + 1);
-                        tokenService.save(token);
-                        if (uuidOptional.isPresent() && uuidOptional.get().equals(token.getToken())) {
-                            TokenStats tokenStats = new TokenStats();
-                            tokenStats.setToken(token);
-
-                            String ipAddress = request.getHeader(IP_HEADER);
-                            if (ipAddress == null) {
-                                tokenStats.setAccessIp(request.getRemoteAddr());
-                            }
-                            tokenStats.setAccessTime(Instant.now());
-                            tokenStats.setResource(request.getMethod() + " " + request.getRequestURI());
-                            tokenStatsService.save(tokenStats);
-                        }
-                    });
-                }
-            }
+            updateTokenStats(body, method, request);
         }
         HttpHeaders httpHeaders = apiProxyService.prepareHttpHeaders(request.getContentType());
         RestTemplate restTemplate = new RestTemplate();
         restTemplate.getMessageConverters().add(0, new StringHttpMessageConverter(StandardCharsets.UTF_8));
         return restTemplate.exchange(uri, method, new HttpEntity<>(body, httpHeaders), String.class).getBody();
     }
 
+
+    @Async
```

Comment:
I also make this method Async. The method should not delay the response of the ApiProxy.  @victoria34